### PR TITLE
PYTHON-1904 Add wrapper classes around libmongocrypt

### PR DESCRIPTION
**Pull request by ShaneHarvey**
This changes makes managing the libmongocrypt CFFI objects safe and idiomatic. It also adds some tests that churn through the mongocrypt state machine using sample input/output data.

~Also included is the `AutoEncrypter` class and `MongoCryptCallback` interface which PyMongo will use to implement the required I/O to mognocryptd, mongod, and kms. I don't plan to add an implementation of `MongoCryptCallback` to this library because I want to avoid any cyclical dependencies: pymongo->pymongocrypt->pymongo~

Edit: I've removed the `AutoEncrypter`/`MongoCryptCallback` classes to simplify this PR. I'll add them back in the next PR.

### CODE REVIEW

File: bindings/python/test/test_mongocrypt.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    def test_mongocrypt_binary_in(self):
+        with MongoCryptBinaryIn(b'1\x0023') as binary:
+            self.assertEqual(binary.to_bytes(), b'1\x0023')
+        self.assertIsNone(binary.bin)
+
+        with MongoCryptBinaryIn(b'') as binary:
+            self.assertEqual(binary.to_bytes(), b'')
+        self.assertIsNone(binary.bin)
+
+        binary = MongoCryptBinaryIn(b'123')
```
Comment (prashantmital) [inline bindings/python/test/test_mongocrypt.py]: Can you clarify what is being tested here with the creation of `binary`? Is it just the use of this type outside of a context manager?

Comment (prashantmital) [inline bindings/python/test/test_mongocrypt.py]: Can you clarify what is being tested here with the creation of `binary`? Is it just the use of this type outside of a context manager?

File: bindings/python/test/test_mongocrypt.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        binary.close()
+        binary.close()
```
Comment (prashantmital) [inline bindings/python/test/test_mongocrypt.py]: Why do we close this twice? Is it to ensure that closing an already closed instance doesn't cause an error?

File: bindings/python/test/test_mongocrypt.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            ({'aws': {'accessKeyId': '', 'secretAccessKey': ''}}, None),
+            ({'aws': {'accessKeyId': 'foo', 'secretAccessKey': 'foo'}}, None),
+            ({'local': {'key': b'1'*96}}, None),
+            ({'aws': {'accessKeyId': 'foo', 'secretAccessKey': 'foo'},
+              'local': {'key': b'1'*96}}, None),
+        ]
+        for kms_providers, schema_map in valid:
+            opts = MongoCryptOptions(kms_providers, schema_map)
+            self.assertEqual(opts.kms_providers, kms_providers)
+            self.assertEqual(opts.schema_map, schema_map)
```
Comment (prashantmital) [inline bindings/python/test/test_mongocrypt.py]: Seems odd that we always use a `NULL`/`None` schema_map. Is it possible to test this with some 'dummy' schema?

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    def to_bytes(self):
+        """Returns this mongocrypt_binary_t as bytes."""
+        data = lib.mongocrypt_binary_data(self.bin)
+        if data == ffi.NULL:
+            return b''
+        data_len = lib.mongocrypt_binary_len(self.bin)
+        return ffi.unpack(ffi.cast("char*", data), data_len)
+
+
+class MongoCryptBinaryOut(_MongoCryptBinary):
```
Comment (prashantmital) [inline bindings/python/pymongocrypt/mongocrypt.py]: Can you add a docstring to indicate that this is the type libmongocrypt returns?

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+
+
+class MongoCryptBinaryIn(_MongoCryptBinary):
+    __slots__ = ("__copy",)
+
+    def __init__(self, data):
+        """Creates a mongocrypt_binary_t from binary data."""
+        # mongocrypt_binary_t does not own the data it is passed so we need to
+        # create a copy to keep the data alive.
+        self.__copy = ffi.new("uint8_t[]", data)
```
Comment (prashantmital) [inline bindings/python/pymongocrypt/mongocrypt.py]: Why use `uint8` here and `char*` on L49? Am I wrong to think that these should have the same type?

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        """Cleanup resources."""
+        if self.__ctx is None:
+            return
+        lib.mongocrypt_ctx_destroy(self.__ctx)
+        self.__ctx = None
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
```
Comment (prashantmital) [inline bindings/python/pymongocrypt/mongocrypt.py]: Is there a reason we aren't using `contextlib`?

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        :Parameters:
+          - `ctx`: A mongocrypt_kms_ctx_t.
+        """
+        self.__ctx = ctx
+
+    def close(self):
+        """Cleanup resources."""
+        self.__ctx = None
+
+    def __enter__(self):
```
Comment (prashantmital) [inline bindings/python/pymongocrypt/mongocrypt.py]: Same comment about contextlib.

File: bindings/python/test/test_mongocrypt.py (hunk 2 above)
Comment (ShaneHarvey) [reply bindings/python/test/test_mongocrypt.py]: I removed this. We always use these types in a with-statement so we don't need `__del__` or `close()`.

File: bindings/python/test/test_mongocrypt.py (hunk 1 above)
Comment (ShaneHarvey) [reply bindings/python/test/test_mongocrypt.py]: I removed this. We always use these types in a with-statement so we don't need `__del__` or `close()`.

File: bindings/python/test/test_mongocrypt.py (hunk 3 above)
Comment (ShaneHarvey) [reply bindings/python/test/test_mongocrypt.py]: Done.

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 6 above)
Comment (prashantmital) [reply bindings/python/pymongocrypt/mongocrypt.py]: My bad. Ignore.

File: bindings/python/test/test_mongocrypt.py (hunk 3 above)
Comment (prashantmital) [reply bindings/python/test/test_mongocrypt.py]: Nice 👍

**Comment by prashantmital** [comment]
LGTM

**Pull request merged**